include/swift/SIL/MemoryLifetime.h

```diff
@@ -0,0 +1,374 @@
+//===--- MemoryLifetime.h ---------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SIL_MEMORY_LIFETIME_H
+#define SWIFT_SIL_MEMORY_LIFETIME_H
+
+#include "swift/SIL/SILBasicBlock.h"
+#include "swift/SIL/SILFunction.h"
+
+namespace swift {
+
+/// A utility for calculating memory lifetime.
+///
+/// The MemoryLifetime utility calculates the lifetime of memory locations.
+/// It provides functions to analyze memory locations and to calculate the
+/// dataflow.
+/// Memory locations are limited to addresses which are guaranteed to be not
+/// aliased, like @in/inout parameters and alloc_stack.
+///
+/// The MemoryLifetime class can be used as base class for optimizations,
```

Comment:
@gottesmm This actually concerns me because I don't want people reaching for it whenever they happen to work with "memory lifetimes". What did you want it for?

Reply:
I want some sort of generally memory lifetime tool. We do this in so many places in the compiler. Why should we reimplement it so much. We need a utility for this!

Reply:
I strongly disagree that this is the right tool for that.

Reply:
@gottesmm you need to specify what you need from a general memory lifetime tool before you can claim that this is the thing. The major source of problems in our code is not the lack of utilities but the misuse and abuse of false abstractions and people latching onto things that have a friendly name without understanding what they need. I honestly have no idea where you would want to use this, but there could be places I haven't thought of.